marlowe-playground-client/src/Marlowe/Linter.purs

```diff
@@ -0,0 +1,160 @@
+module Marlowe.Linter
+  ( lint
+  , State(..)
+  , Position
+  , _holes
+  , _negativeDeposits
+  , _negativePayments
+  ) where
+
+import Prelude
+import Data.Array (catMaybes, cons, fold, (:))
+import Data.BigInteger (BigInteger)
+import Data.Lens (Lens', over, set, view)
+import Data.Lens.Iso.Newtype (_Newtype)
+import Data.Lens.Record (prop)
+import Data.Maybe (Maybe(..))
+import Data.Newtype (class Newtype)
+import Data.Symbol (SProxy(..))
+import Data.Tuple.Nested (type (/\), get1, get2, get3, (/\))
+import Marlowe.Holes (Action(..), Case(..), Contract(..), Holes, Term(..), Value(..), getHoles, insertHole)
+import Text.Parsing.StringParser (Pos)
+
+type Position
+  = { start :: Pos, end :: Pos }
+
+newtype State
+  = State
+  { holes :: Holes
+  , negativePayments :: Array Position
+  , negativeDeposits :: Array Position
+  }
+
+derive instance newtypeState :: Newtype State _
+
+instance semigroupState :: Semigroup State where
+  append (State s1) (State s2) =
+    State
+      { holes: s1.holes <> s2.holes
+      , negativePayments: s1.negativePayments <> s2.negativePayments
+      , negativeDeposits: s1.negativeDeposits <> s2.negativeDeposits
+      }
+
+instance monoidState :: Monoid State where
+  mempty =
+    State
+      { holes: mempty
+      , negativePayments: mempty
+      , negativeDeposits: mempty
+      }
+
+_holes :: Lens' State Holes
+_holes = _Newtype <<< prop (SProxy :: SProxy "holes")
+
+_negativePayments :: Lens' State (Array Position)
+_negativePayments = _Newtype <<< prop (SProxy :: SProxy "negativePayments")
+
+_negativeDeposits :: Lens' State (Array Position)
+_negativeDeposits = _Newtype <<< prop (SProxy :: SProxy "negativeDeposits")
+
+lint :: Term Contract -> State
+lint = traverseContract mempty
+
+-- | We go through a contract term collecting all warnings and holes etc so that we can display them in the editor
+traverseContract :: State -> Term Contract -> State
+traverseContract state (Term Close _ _) = state
+
+traverseContract state (Term (Pay acc payee token payment contract) start end) =
+  let
+    hs = view _holes state
+
+    holes = getHoles hs acc <> getHoles hs payee <> getHoles hs token <> getHoles hs payment
+
+    stateWithHoles = set _holes holes state
```

Comment:
Hmm...tell me if I'm misunderstanding the code, but don't you want to use the `getHoles` instance for `Contract` through this code? Looks like you're repeating the boilerplate traversal stuff...

Reply:
I'm repeating the boilerplate stuff to avoid traversing the contract more than is necessary. I want this code to be as performant as possible since it runs every time the contract is changed in the editor. For this reason I want to traverse the contract only once, collecting all warnings (holes etc) as I go.

I should remove unused instances in this PR to avoid confusion.

Reply:
Oh, is the significant difference here that, for instance, we don't look at the contract in `Pay _ _ _ _ contract`?

Reply:
exactly, we traverse that with the recursive call to `traverseContract` rather than `getHoles` since we also want to find negative payments etc. I could get rid of `HasMarlowHoles` class if it makes things confusing?

Reply:
Yeah, if it's otherwise unused I think that would help. :-)

Reply:
ok, big refactoring pushed. I've left `HasMarloweHoles` in because removing it would be a lot of work, however I have removed all unused instances. Now we have a similar pattern through all of the cases:
```
let
  holes = something
  negativePayments = something
  newState = something
in
  traverseContract newState contract
```
Hopefully this makes it easier to read, especially because we will be adding more things to the `State` in the future as we come up with new types of warning.

Reply:
This will also be important in the future because we will have things in `State` that the next step in the traversal need to know about. Previously I was passing `mempty` into the next recursion but this will lose track of any state we might need. Specifically in the future we will be checking that timeouts always increase so we will need to keep track of the current max timeout in each step in the traversal.